---
title: "Uniform Interface for Genomic Variation Signature Analysis and Visualization"
author: "Shixiang Wang \\

        ShanghaiTech. University"
date: "`r Sys.Date()`"
output:
  prettydoc::html_pretty:
    toc: true
    theme: cayman
    highlight: github
  pdf_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{Usage documentation}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Overview

The goal of **sigminer** is to provide an uniform interface for genomic variation signature analysis and visualization. **sigminer** is originated from [VSHunter](https://github.com/ShixiangWang/VSHunter) package I wrote. I hate ugly structure and function names in VSHunter, thus reconstruct it using concise function names, S4 classes and S3 methods etc.. I will continue to add more features to uncover genomic variation signatures and their correlationship with phenotypes and genotypes.

**sigminer** is powered by [NMF](https://github.com/renozao/NMF) package and [maftools](https://github.com/PoisonAlien/maftools) package.

## Installation

You can install the development version of sigminer from Github with:

```{r, eval=FALSE}
remotes::install_github("ShixiangWang/sigminer")
```


## Citation

If you use **sigminer** in academic field, please cite:

```
Wang, Shixiang, et al. “APOBEC3B and APOBEC mutational signature as
    potential predictive markers for immunotherapy response in non-small
    cell lung cancer.” Oncogene (2018).
```

and

```
Gaujoux, Renaud, and Cathal Seoighe. "A Flexible R Package for 
    Nonnegative Matrix Factorization."" BMC Bioinformatics 11, no. 1 (December 2010).
```

## Acknowledgments

The code for extracting copy number signatures was based in part on the source code 
from paper *Copy number signatures and mutational processes in ovarian carcinoma*, 
if you use this feature, please also cite:

```
Macintyre, Geoff, et al. "Copy number signatures and mutational
    processes in ovarian carcinoma." Nature genetics 50.9 (2018): 1262.
```

The code for extracting mutational signatures was based in part on the source 
code of the **maftools** package, if you use this feature, please also cite:

```
Mayakonda, Anand, et al. "Maftools: efficient and comprehensive analysis
    of somatic variants in cancer." Genome research 28.11 (2018): 1747-1756.
```


## LICENSE

MIT © Shixiang Wang
