<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Extract, Analyze and Visualize Signatures for Genomic
    Variations • sigminer</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Extract, Analyze and Visualize Signatures for Genomic
    Variations">
<meta property="og:description" content="Genomic alterations including single nucleotide
    substitution, copy number alteration, etc. are the major force for
    cancer initialization and development. Due to the specificity of
    molecular lesions caused by genomic alterations, we can generate
    characteristic alteration spectra, called signature (Wang, Shixiang,
    et al.  (2020) &lt;DOI:10.1101/2020.04.27.20082404&gt; &amp; Alexandrov, Ludmil
    B., et al. (2020) &lt;DOI:10.1038/s41586-020-1943-3&gt; &amp; Macintyre, Geoff,
    et al. (2018) &lt;DOI:10.1038/s41588-018-0179-8&gt;).  This package helps
    users to extract, analyze and visualize signatures from genomic
    alteration records, thus providing new insight into cancer study.">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">sigminer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.13</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/sigminer.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ShixiangWang/sigminer/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="sigminer-mutational-signature-analysis-and-visualization-in-r-" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#sigminer-mutational-signature-analysis-and-visualization-in-r-" class="anchor"></a>Sigminer: Mutational Signature Analysis and Visualization in R <img src="reference/figures/logo.png" align="right" height="140" width="120" alt="logo">
</h1></div>

<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>Genomic alterations including single nucleotide substitution (SBS), copy number alteration (CNA), etc. are the major force for cancer initialization and development. Due to the specificity of molecular lesions caused by genomic alterations, we can generate characteristic alteration spectra, called ‘signature’. This package helps users to extract, analyze and visualize signatures from genomic alteration records, thus providing new insight into cancer study.</p>
<p><strong>SBS signatures</strong>:</p>
<p><img src="reference/figures/README-unnamed-chunk-1-1.png" width="100%"></p>
<p><strong>Copy number signatures</strong>:</p>
<p><img src="reference/figures/README-unnamed-chunk-2-1.png" width="100%"></p>
<p><strong>DBS signatures</strong>:</p>
<p><img src="reference/figures/README-unnamed-chunk-3-1.png" width="100%"></p>
<p><strong>INDEL (i.e. ID) signatures</strong>:</p>
<p><img src="reference/figures/README-unnamed-chunk-4-1.png" width="100%"></p>
<div id="feature" class="section level3">
<h3 class="hasAnchor">
<a href="#feature" class="anchor"></a>Feature</h3>
<ul>
<li>supports a standard <em>de novo</em> pipeline for identification of <strong>4</strong> types of signatures: copy number, SBS, DBS and INDEL</li>
<li>supports quantify exposure for one sample based on <em>known signatures</em>
</li>
<li>supports two methods for calling copy number signatures: one is from Macintyre et al. 2018 and the other is created by us</li>
<li>supports association and group analysis and visualization for signatures</li>
<li>supports a bayesian variant of NMF algorithm to enable optimal inferences for the number of signatures through the automatic relevance determination technique from <strong>SignatureAnalyzer</strong> package</li>
<li>supports two plot styles for signature profile: ‘default’ (like <strong>SignatureAnalyzer</strong> package) and ‘cosmic’ (like <strong>COSMIC</strong> database)</li>
<li>supports two types of signatrue exposures: relative exposure (relative contribution of signatures in each sample) and absolute exposure (estimated variation records of signatures in each sample)</li>
<li>supports basic summary and visualization for profile of mutation (powered by <strong>maftools</strong>) and copy number</li>
<li>supports parallel computation by R packages <strong>foreach</strong>, <strong>future</strong> and <strong>NMF</strong>
</li>
<li>efficient code powered by R packages <strong>data.table</strong> and <strong>tidyverse</strong>
</li>
<li>elegant plots powered by R packages <strong>ggplot2</strong>, <strong>ggpubr</strong>, <strong>cowplot</strong> and <strong>patchwork</strong>
</li>
<li>well tested by R package <strong>testthat</strong> and documented by R package <strong>roxygen2</strong>, <strong>roxytest</strong>, <strong>pkgdown</strong>, and etc. for both reliable and reproducible research</li>
</ul>
</div>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install the stable release of <strong>sigminer</strong> from CRAN with:</p>
<div class="sourceCode" id="cb1"><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"sigminer"</span>, <span class="kw">dependencies</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="co"># Or</span>
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"sigminer"</span>, <span class="kw">dependencies</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></div>
<p>You can install the development version of <strong>sigminer</strong> from Github with:</p>
<div class="sourceCode" id="cb2"><pre class="r"><span class="kw pkg">remotes</span><span class="kw ns">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span>(<span class="st">"ShixiangWang/sigminer"</span>, <span class="kw">dependencies</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="co"># For Chinese users, run </span>
<span class="kw pkg">remotes</span><span class="kw ns">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_git.html">install_git</a></span>(<span class="st">"https://gitee.com/ShixiangWang/sigminer"</span>, <span class="kw">dependencies</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p>A complete documentation of <strong>sigminer</strong> can be read online at <a href="https://shixiangwang.github.io/sigminer-doc/" class="uri">https://shixiangwang.github.io/sigminer-doc/</a> (For Chinese users, you can also read it at <a href="https://shixiangwang.gitee.io/sigminer-doc" class="uri">https://shixiangwang.gitee.io/sigminer-doc</a>). All functions are well organized and documented at <a href="https://shixiangwang.github.io/sigminer/reference/index.html" class="uri">https://shixiangwang.github.io/sigminer/reference/index.html</a> (For Chinese users, you can also read it at <a href="https://shixiangwang.gitee.io/sigminer/reference/index.html" class="uri">https://shixiangwang.gitee.io/sigminer/reference/index.html</a>). For usage of a specific function <code>fun</code>, run <code>?fun</code> in your R console to see its documentation.</p>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<hr>
<p><strong><em>Wang, Shixiang, et al. “Copy number signature analyses in prostate cancer reveal distinct etiologies and clinical outcomes” medRxiv (2020)</em></strong> <a href="https://www.medrxiv.org/content/early/2020/04/29/2020.04.27.20082404" class="uri">https://www.medrxiv.org/content/early/2020/04/29/2020.04.27.20082404</a></p>
<hr>
</div>
<div id="download-stats" class="section level2">
<h2 class="hasAnchor">
<a href="#download-stats" class="anchor"></a>Download Stats</h2>
<p><img src="reference/figures/README-unnamed-chunk-7-1.png" width="100%"></p>
</div>
<div id="acknowledgments" class="section level2">
<h2 class="hasAnchor">
<a href="#acknowledgments" class="anchor"></a>Acknowledgments</h2>
<p>If you use <strong>NMF</strong> package in R, please also cite:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>Gaujoux, Renaud, and Cathal Seoighe. <span class="st">"A Flexible R Package for </span></span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="st">    Nonnegative Matrix Factorization."" BMC Bioinformatics 11, no. 1 (December 2010).</span></span></code></pre></div>
<p>The method “M” for extracting copy number signatures was based in part on the source code from paper <em>Copy number signatures and mutational processes in ovarian carcinoma</em>, if you use this feature, please also cite:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>Macintyre, Geoff, et al. <span class="st">"Copy number signatures and mutational</span></span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="st">    processes in ovarian carcinoma."</span> Nature genetics <span class="fl">50.9</span> (<span class="dv">2018</span>)<span class="op">:</span><span class="st"> </span><span class="fl">1262.</span></span></code></pre></div>
<p>The code for extracting SBS signatures was based in part on the source code of the <strong>maftools</strong> package, if you use this feature, please also cite:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>Mayakonda, Anand, et al. <span class="st">"Maftools: efficient and comprehensive analysis</span></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="st">    of somatic variants in cancer."</span> Genome research <span class="fl">28.11</span> (<span class="dv">2018</span>)<span class="op">:</span><span class="st"> </span><span class="dv">1747</span><span class="fl">-1756.</span></span></code></pre></div>
<p>The code for extracting mutational signatures was based in part on the source code of the <strong>SignatureAnalyzer</strong> package, if you use this feature, please also cite:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>Kim, Jaegil, et al. <span class="st">"Somatic ERCC2 mutations are associated with a distinct genomic</span></span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="st">    signature in urothelial tumors."</span> Nature genetics <span class="fl">48.6</span> (<span class="dv">2016</span>)<span class="op">:</span><span class="st"> </span><span class="fl">600.</span></span></code></pre></div>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<ol>
<li>Alexandrov, Ludmil B., et al. “The repertoire of mutational signatures in human cancer.” Nature 578.7793 (2020): 94-101.</li>
<li>Macintyre, Geoff, et al. “Copy number signatures and mutational processes in ovarian carcinoma.” Nature genetics 50.9 (2018): 1262.</li>
<li>Mayakonda, Anand, et al. “Maftools: efficient and comprehensive analysis of somatic variants in cancer.” Genome research 28.11 (2018): 1747-1756.</li>
<li>Gaujoux, Renaud, and Cathal Seoighe. “A Flexible R Package for Nonnegative Matrix Factorization.”" BMC Bioinformatics 11, no. 1 (December 2010).</li>
<li>H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2016.</li>
<li>Tan, Vincent YF, and Cédric Févotte. “Automatic relevance determination in nonnegative matrix factorization with the/spl beta/-divergence.” IEEE Transactions on Pattern Analysis and Machine Intelligence 35.7 (2012): 1592-1605.</li>
<li>Kim, Jaegil, et al. “Somatic ERCC2 mutations are associated with a distinct genomic signature in urothelial tumors.” Nature genetics 48.6 (2016): 600.</li>
<li>Bergstrom EN, Huang MN, Mahto U, Barnes M, Stratton MR, Rozen SG, Alexandrov LB: SigProfilerMatrixGenerator: a tool for visualizing and exploring patterns of small mutational events. BMC Genomics 2019, 20:685 <a href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-019-6041-2" class="uri">https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-019-6041-2</a>
</li>
</ol>
</div>
<div id="license" class="section level2">
<h2 class="hasAnchor">
<a href="#license" class="anchor"></a>LICENSE</h2>
<p>The software is made available for non commercial research purposes only under the <a href="https://github.com/ShixiangWang/sigminer/blob/master/LICENSE.md">MIT</a>. However, notwithstanding any provision of the MIT License, the software currently may not be used for commercial purposes without explicit written permission after contacting Shixiang Wang <a href="mailto:wangshx@shanghaitech.edu.cn" class="email">wangshx@shanghaitech.edu.cn</a> or Xue-Song Liu <a href="mailto:liuxs@shanghaitech.edu.cn" class="email">liuxs@shanghaitech.edu.cn</a>.</p>
<p>MIT © 2019-2020 Shixiang Wang, Xue-Song Liu</p>
<p>MIT © 2018 Geoffrey Macintyre</p>
<p>MIT © 2018 Anand Mayakonda</p>
<hr>
<p><strong><a href="https://github.com/XSLiuLab">Cancer Biology Group</a> @ShanghaiTech</strong></p>
<p><strong>Research group led by Xue-Song Liu in ShanghaiTech University</strong></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from CRAN at <br><a href="https://cloud.r-project.org/package=sigminer">https://​cloud.r-project.org/​package=sigminer</a>
</li>
<li>Browse source code at <br><a href="https://github.com/ShixiangWang/sigminer/">https://​github.com/​ShixiangWang/​sigminer/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/ShixiangWang/sigminer/issues">https://​github.com/​ShixiangWang/​sigminer/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing sigminer</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://orcid.org/0000-0001-9855-7357">Shixiang Wang</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-9855-7357" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://orcid.org/0000-0002-7736-0077">Xue-Song Liu</a> <br><small class="roles"> Author, contributor </small> <a href="https://orcid.org/0000-0002-7736-0077" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Tao Wu <br><small class="roles"> Author, contributor </small>  </li>
<li>Ziyu Tao <br><small class="roles"> Author, contributor </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://cran.r-project.org/package=sigminer"><img src="https://www.r-pkg.org/badges/version/sigminer" alt="CRAN status"></a></li>
<li><a href="https://www.tidyverse.org/lifecycle/#stable"><img src="https://img.shields.io/badge/lifecycle-stable-blue.svg" alt="lifecycle"></a></li>
<li><a href="https://codecov.io/github/ShixiangWang/sigminer?branch=master"><img src="https://codecov.io/gh/ShixiangWang/sigminer/branch/master/graph/badge.svg" alt="Coverage status"></a></li>
<li><a href="https://cran.r-project.org/package=sigminer"><img src="http://cranlogs.r-pkg.org/badges/grand-total/sigminer?color=orange"></a></li>
<li><a href="https://github.com/ShixiangWang/sigminer/issues?q=is%3Aissue+is%3Aclosed"><img src="https://img.shields.io/github/issues-closed/ShixiangWang/sigminer.svg" alt="Closed issues"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://orcid.org/0000-0001-9855-7357">Shixiang Wang</a>, <a href="https://orcid.org/0000-0002-7736-0077">Xue-Song Liu</a>, Tao Wu, Ziyu Tao.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
