<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Uniform Interface for Genomic Variation Signature Analysis and Visualization • sigminer</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Uniform Interface for Genomic Variation Signature Analysis and Visualization">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sigminer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/sigminer.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ShixiangWang/sigminer">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Uniform Interface for Genomic Variation Signature Analysis and Visualization</h1>
                        <h4 class="author">Shixiang Wang<br>
ShanghaiTech. University</h4>
            
            <h4 class="date">2019-04-25</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ShixiangWang/sigminer/blob/master/vignettes/sigminer.Rmd"><code>vignettes/sigminer.Rmd</code></a></small>
      <div class="hidden name"><code>sigminer.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>The goal of <strong>sigminer</strong> is to provide an uniform interface for genomic variation signature analysis and visualization. <strong>sigminer</strong> is originated from <a href="https://github.com/ShixiangWang/VSHunter">VSHunter</a> package I wrote. I hate ugly structure and function names in VSHunter, thus reconstruct it using concise function names, S4 classes and S3 methods etc.. I will continue to add more features to uncover genomic variation signatures and their correlationship with phenotypes and genotypes.</p>
<p><strong>sigminer</strong> is powered by <a href="https://github.com/renozao/NMF">NMF</a> package and <a href="https://github.com/PoisonAlien/maftools">maftools</a> package.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install the stable release of <strong>sigminer</strong> from CRAN with:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"sigminer"</span>)</a></code></pre></div>
<p>You can also install the development version of sigminer from Github with:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">remotes<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/remotes/topics/install_github">install_github</a></span>(<span class="st">"ShixiangWang/sigminer"</span>)</a></code></pre></div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p>An example for how to extract mutational signatures are given as the following.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(sigminer)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="co">#&gt; Registered S3 methods overwritten by 'ggplot2':</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="co">#&gt;   method         from </span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="co">#&gt;   [.quosures     rlang</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="co">#&gt;   c.quosures     rlang</span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="co">#&gt;   print.quosures rlang</span></a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="co">#&gt; Welcome to 'sigminer' package!</span></a>
<a class="sourceLine" id="cb3-8" data-line-number="8"><span class="co">#&gt; ======================================================</span></a>
<a class="sourceLine" id="cb3-9" data-line-number="9"><span class="co">#&gt; sigminer version 0.1.8</span></a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="co">#&gt; Github page: https://github.com/ShixiangWang/sigminer</span></a>
<a class="sourceLine" id="cb3-11" data-line-number="11"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb3-12" data-line-number="12"><span class="co">#&gt; More info please call 'hello()' in console.</span></a>
<a class="sourceLine" id="cb3-13" data-line-number="13"><span class="co">#&gt; ======================================================</span></a>
<a class="sourceLine" id="cb3-14" data-line-number="14"><span class="co">#&gt; </span></a></code></pre></div>
<div id="load-data-as-a-maf-object" class="section level3">
<h3 class="hasAnchor">
<a href="#load-data-as-a-maf-object" class="anchor"></a>Load data as a MAF object</h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">laml.maf &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"tcga_laml.maf.gz"</span>, <span class="dt">package =</span> <span class="st">"maftools"</span>)</a>
<a class="sourceLine" id="cb4-2" data-line-number="2">laml &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_maf.html">read_maf</a></span>(<span class="dt">maf =</span> laml.maf)</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="co">#&gt; reading maf..</span></a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="co">#&gt; silent variants: 475</span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="co">#&gt;         ID   N</span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="co">#&gt; 1: Samples 157</span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7"><span class="co">#&gt; 2: 5'Flank   3</span></a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="co">#&gt; 3:     IGR   5</span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9"><span class="co">#&gt; 4:  Intron   8</span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10"><span class="co">#&gt; 5:     RNA  10</span></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="co">#&gt; 6:  Silent 449</span></a>
<a class="sourceLine" id="cb4-12" data-line-number="12"><span class="co">#&gt; Summarizing..</span></a>
<a class="sourceLine" id="cb4-13" data-line-number="13"><span class="co">#&gt;                    ID          summary  Mean Median</span></a>
<a class="sourceLine" id="cb4-14" data-line-number="14"><span class="co">#&gt;  1:        NCBI_Build               37    NA     NA</span></a>
<a class="sourceLine" id="cb4-15" data-line-number="15"><span class="co">#&gt;  2:            Center genome.wustl.edu    NA     NA</span></a>
<a class="sourceLine" id="cb4-16" data-line-number="16"><span class="co">#&gt;  3:           Samples              193    NA     NA</span></a>
<a class="sourceLine" id="cb4-17" data-line-number="17"><span class="co">#&gt;  4:            nGenes             1241    NA     NA</span></a>
<a class="sourceLine" id="cb4-18" data-line-number="18"><span class="co">#&gt;  5:   Frame_Shift_Del               52 0.271      0</span></a>
<a class="sourceLine" id="cb4-19" data-line-number="19"><span class="co">#&gt;  6:   Frame_Shift_Ins               91 0.474      0</span></a>
<a class="sourceLine" id="cb4-20" data-line-number="20"><span class="co">#&gt;  7:      In_Frame_Del               10 0.052      0</span></a>
<a class="sourceLine" id="cb4-21" data-line-number="21"><span class="co">#&gt;  8:      In_Frame_Ins               42 0.219      0</span></a>
<a class="sourceLine" id="cb4-22" data-line-number="22"><span class="co">#&gt;  9: Missense_Mutation             1342 6.990      7</span></a>
<a class="sourceLine" id="cb4-23" data-line-number="23"><span class="co">#&gt; 10: Nonsense_Mutation              103 0.536      0</span></a>
<a class="sourceLine" id="cb4-24" data-line-number="24"><span class="co">#&gt; 11:       Splice_Site               92 0.479      0</span></a>
<a class="sourceLine" id="cb4-25" data-line-number="25"><span class="co">#&gt; 12:             total             1732 9.021      9</span></a>
<a class="sourceLine" id="cb4-26" data-line-number="26"><span class="co">#&gt; Gene Summary..</span></a>
<a class="sourceLine" id="cb4-27" data-line-number="27"><span class="co">#&gt;       Hugo_Symbol Frame_Shift_Del Frame_Shift_Ins In_Frame_Del</span></a>
<a class="sourceLine" id="cb4-28" data-line-number="28"><span class="co">#&gt;    1:        FLT3               0               0            1</span></a>
<a class="sourceLine" id="cb4-29" data-line-number="29"><span class="co">#&gt;    2:      DNMT3A               4               0            0</span></a>
<a class="sourceLine" id="cb4-30" data-line-number="30"><span class="co">#&gt;    3:        NPM1               0              33            0</span></a>
<a class="sourceLine" id="cb4-31" data-line-number="31"><span class="co">#&gt;    4:        IDH2               0               0            0</span></a>
<a class="sourceLine" id="cb4-32" data-line-number="32"><span class="co">#&gt;    5:        IDH1               0               0            0</span></a>
<a class="sourceLine" id="cb4-33" data-line-number="33"><span class="co">#&gt;   ---                                                         </span></a>
<a class="sourceLine" id="cb4-34" data-line-number="34"><span class="co">#&gt; 1237:      ZNF689               0               0            0</span></a>
<a class="sourceLine" id="cb4-35" data-line-number="35"><span class="co">#&gt; 1238:      ZNF75D               0               0            0</span></a>
<a class="sourceLine" id="cb4-36" data-line-number="36"><span class="co">#&gt; 1239:      ZNF827               1               0            0</span></a>
<a class="sourceLine" id="cb4-37" data-line-number="37"><span class="co">#&gt; 1240:       ZNF99               0               0            0</span></a>
<a class="sourceLine" id="cb4-38" data-line-number="38"><span class="co">#&gt; 1241:        ZPBP               0               0            0</span></a>
<a class="sourceLine" id="cb4-39" data-line-number="39"><span class="co">#&gt;       In_Frame_Ins Missense_Mutation Nonsense_Mutation Splice_Site total</span></a>
<a class="sourceLine" id="cb4-40" data-line-number="40"><span class="co">#&gt;    1:           33                15                 0           3    52</span></a>
<a class="sourceLine" id="cb4-41" data-line-number="41"><span class="co">#&gt;    2:            0                39                 5           6    54</span></a>
<a class="sourceLine" id="cb4-42" data-line-number="42"><span class="co">#&gt;    3:            0                 1                 0           0    34</span></a>
<a class="sourceLine" id="cb4-43" data-line-number="43"><span class="co">#&gt;    4:            0                20                 0           0    20</span></a>
<a class="sourceLine" id="cb4-44" data-line-number="44"><span class="co">#&gt;    5:            0                18                 0           0    18</span></a>
<a class="sourceLine" id="cb4-45" data-line-number="45"><span class="co">#&gt;   ---                                                                   </span></a>
<a class="sourceLine" id="cb4-46" data-line-number="46"><span class="co">#&gt; 1237:            0                 1                 0           0     1</span></a>
<a class="sourceLine" id="cb4-47" data-line-number="47"><span class="co">#&gt; 1238:            0                 1                 0           0     1</span></a>
<a class="sourceLine" id="cb4-48" data-line-number="48"><span class="co">#&gt; 1239:            0                 0                 0           0     1</span></a>
<a class="sourceLine" id="cb4-49" data-line-number="49"><span class="co">#&gt; 1240:            0                 1                 0           0     1</span></a>
<a class="sourceLine" id="cb4-50" data-line-number="50"><span class="co">#&gt; 1241:            0                 1                 0           0     1</span></a>
<a class="sourceLine" id="cb4-51" data-line-number="51"><span class="co">#&gt;       MutatedSamples AlteredSamples</span></a>
<a class="sourceLine" id="cb4-52" data-line-number="52"><span class="co">#&gt;    1:             52             52</span></a>
<a class="sourceLine" id="cb4-53" data-line-number="53"><span class="co">#&gt;    2:             48             48</span></a>
<a class="sourceLine" id="cb4-54" data-line-number="54"><span class="co">#&gt;    3:             33             33</span></a>
<a class="sourceLine" id="cb4-55" data-line-number="55"><span class="co">#&gt;    4:             20             20</span></a>
<a class="sourceLine" id="cb4-56" data-line-number="56"><span class="co">#&gt;    5:             18             18</span></a>
<a class="sourceLine" id="cb4-57" data-line-number="57"><span class="co">#&gt;   ---                              </span></a>
<a class="sourceLine" id="cb4-58" data-line-number="58"><span class="co">#&gt; 1237:              1              1</span></a>
<a class="sourceLine" id="cb4-59" data-line-number="59"><span class="co">#&gt; 1238:              1              1</span></a>
<a class="sourceLine" id="cb4-60" data-line-number="60"><span class="co">#&gt; 1239:              1              1</span></a>
<a class="sourceLine" id="cb4-61" data-line-number="61"><span class="co">#&gt; 1240:              1              1</span></a>
<a class="sourceLine" id="cb4-62" data-line-number="62"><span class="co">#&gt; 1241:              1              1</span></a>
<a class="sourceLine" id="cb4-63" data-line-number="63"><span class="co">#&gt; Checking clinical data..</span></a>
<a class="sourceLine" id="cb4-64" data-line-number="64"><span class="co">#&gt; </span><span class="al">NOTE</span><span class="co">: Missing clinical data! It is strongly recommended to provide clinical data associated with samples if available.</span></a>
<a class="sourceLine" id="cb4-65" data-line-number="65"><span class="co">#&gt; Done !</span></a></code></pre></div>
</div>
<div id="prepare-data-for-signature-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#prepare-data-for-signature-analysis" class="anchor"></a>Prepare data for signature analysis</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(BSgenome.Hsapiens.UCSC.hg19, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="co">#&gt; 载入程辑包：'S4Vectors'</span></a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="co">#&gt; The following object is masked from 'package:base':</span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="co">#&gt;     expand.grid</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="co">#&gt; 载入程辑包：'Biostrings'</span></a>
<a class="sourceLine" id="cb5-9" data-line-number="9"><span class="co">#&gt; The following object is masked from 'package:base':</span></a>
<a class="sourceLine" id="cb5-10" data-line-number="10"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb5-11" data-line-number="11"><span class="co">#&gt;     strsplit</span></a>
<a class="sourceLine" id="cb5-12" data-line-number="12"></a>
<a class="sourceLine" id="cb5-13" data-line-number="13">sig_pre &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sig_prepare.html">sig_prepare</a></span>(laml, <span class="dt">ref_genome =</span> <span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span>, </a>
<a class="sourceLine" id="cb5-14" data-line-number="14">                      <span class="dt">prefix =</span> <span class="st">"chr"</span>, <span class="dt">add =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb5-15" data-line-number="15"><span class="co">#&gt; Warning in maftools::trinucleotideMatrix(maf, ref_genome = ref_genome, prefix = prefix, : Chromosome names in MAF must match chromosome names in reference genome.</span></a>
<a class="sourceLine" id="cb5-16" data-line-number="16"><span class="co">#&gt; Ignorinig 101 single nucleotide variants from missing chromosomes chr23</span></a>
<a class="sourceLine" id="cb5-17" data-line-number="17"><span class="co">#&gt; Extracting 5' and 3' adjacent bases..</span></a>
<a class="sourceLine" id="cb5-18" data-line-number="18"><span class="co">#&gt; Extracting +/- 20bp around mutated bases for background C&gt;T estimation..</span></a>
<a class="sourceLine" id="cb5-19" data-line-number="19"><span class="co">#&gt; Estimating APOBEC enrichment scores..</span></a>
<a class="sourceLine" id="cb5-20" data-line-number="20"><span class="co">#&gt; Performing one-way Fisher's test for APOBEC enrichment..</span></a>
<a class="sourceLine" id="cb5-21" data-line-number="21"><span class="co">#&gt; APOBEC related mutations are enriched in 3.315% of samples (APOBEC enrichment score &gt; 2 ; 6 of 181 samples)</span></a>
<a class="sourceLine" id="cb5-22" data-line-number="22"><span class="co">#&gt; Creating mutation matrix..</span></a>
<a class="sourceLine" id="cb5-23" data-line-number="23"><span class="co">#&gt; matrix of dimension 188x96</span></a></code></pre></div>
</div>
<div id="extract-signatures" class="section level3">
<h3 class="hasAnchor">
<a href="#extract-signatures" class="anchor"></a>Extract signatures</h3>
<p>Before extracting signatures, we can estimate signature number with <code>sig_estimate</code> function.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(NMF)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="co">#&gt; 载入需要的程辑包：pkgmaker</span></a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="co">#&gt; 载入需要的程辑包：registry</span></a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5"><span class="co">#&gt; 载入程辑包：'pkgmaker'</span></a>
<a class="sourceLine" id="cb6-6" data-line-number="6"><span class="co">#&gt; The following object is masked from 'package:S4Vectors':</span></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb6-8" data-line-number="8"><span class="co">#&gt;     new2</span></a>
<a class="sourceLine" id="cb6-9" data-line-number="9"><span class="co">#&gt; The following object is masked from 'package:base':</span></a>
<a class="sourceLine" id="cb6-10" data-line-number="10"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb6-11" data-line-number="11"><span class="co">#&gt;     isFALSE</span></a>
<a class="sourceLine" id="cb6-12" data-line-number="12"><span class="co">#&gt; 载入需要的程辑包：rngtools</span></a>
<a class="sourceLine" id="cb6-13" data-line-number="13"><span class="co">#&gt; 载入需要的程辑包：cluster</span></a>
<a class="sourceLine" id="cb6-14" data-line-number="14"><span class="co">#&gt; NMF - BioConductor layer [OK] | Shared memory capabilities [NO: bigmemory] | Cores 7/8</span></a>
<a class="sourceLine" id="cb6-15" data-line-number="15"><span class="co">#&gt;   To enable shared memory capabilities, try: install.extras('</span></a>
<a class="sourceLine" id="cb6-16" data-line-number="16"><span class="co">#&gt; NMF</span></a>
<a class="sourceLine" id="cb6-17" data-line-number="17"><span class="co">#&gt; ')</span></a>
<a class="sourceLine" id="cb6-18" data-line-number="18"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb6-19" data-line-number="19"><span class="co">#&gt; 载入程辑包：'NMF'</span></a>
<a class="sourceLine" id="cb6-20" data-line-number="20"><span class="co">#&gt; The following object is masked from 'package:S4Vectors':</span></a>
<a class="sourceLine" id="cb6-21" data-line-number="21"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb6-22" data-line-number="22"><span class="co">#&gt;     nrun</span></a></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">sig_est &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sig_estimate.html">sig_estimate</a></span>(sig_pre<span class="op">$</span>nmf_matrix, <span class="dt">range =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">5</span>, <span class="dt">pConstant =</span> <span class="fl">0.01</span>)</a></code></pre></div>
<p><img src="sigminer_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p><code>sig_est</code> keeps all information of estimation.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">sig_laml &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sig_extract.html">sig_extract</a></span>(sig_pre<span class="op">$</span>nmf_matrix, <span class="dt">n_sig =</span> <span class="dv">2</span>, <span class="dt">mode =</span> <span class="st">"mutation"</span>, <span class="dt">pConstant =</span> <span class="fl">0.01</span>)</a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="co">#&gt; Comparing against experimentally validated 30 signatures.. (See http://cancer.sanger.ac.uk/cosmic/signatures for details.)</span></a>
<a class="sourceLine" id="cb8-3" data-line-number="3"><span class="co">#&gt; Found Signature_1 most similar to validated Signature_1. Aetiology: spontaneous deamination of 5-methylcytosine [cosine-similarity: 0.933]</span></a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="co">#&gt; Found Signature_2 most similar to validated Signature_1. Aetiology: spontaneous deamination of 5-methylcytosine [cosine-similarity: 0.846]</span></a></code></pre></div>
<blockquote>
<p>Please note the ‘pConstant’ option:<br>
It is a small positive value to add to the matrix. Use it ONLY if the functions throws an non-conformable arrays error.</p>
</blockquote>
</div>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<p>If you use <strong>sigminer</strong> in academic field, please cite:</p>
<pre><code>Wang, Shixiang, et al. “APOBEC3B and APOBEC mutational signature as
    potential predictive markers for immunotherapy response in non-small
    cell lung cancer.” Oncogene (2018).</code></pre>
<p>and</p>
<pre><code>Gaujoux, Renaud, and Cathal Seoighe. "A Flexible R Package for 
    Nonnegative Matrix Factorization."" BMC Bioinformatics 11, no. 1 (December 2010).</code></pre>
</div>
<div id="acknowledgments" class="section level2">
<h2 class="hasAnchor">
<a href="#acknowledgments" class="anchor"></a>Acknowledgments</h2>
<p>The code for extracting copy number signatures was based in part on the source code from paper <em>Copy number signatures and mutational processes in ovarian carcinoma</em>, if you use this feature, please also cite:</p>
<pre><code>Macintyre, Geoff, et al. "Copy number signatures and mutational
    processes in ovarian carcinoma." Nature genetics 50.9 (2018): 1262.</code></pre>
<p>The code for extracting mutational signatures was based in part on the source code of the <strong>maftools</strong> package, if you use this feature, please also cite:</p>
<pre><code>Mayakonda, Anand, et al. "Maftools: efficient and comprehensive analysis
    of somatic variants in cancer." Genome research 28.11 (2018): 1747-1756.</code></pre>
</div>
<div id="license" class="section level2">
<h2 class="hasAnchor">
<a href="#license" class="anchor"></a>LICENSE</h2>
<p>MIT © 2019 Shixiang Wang, Geoffrey Macintyre, Xue-Song Liu</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#overview">Overview</a></li>
      <li><a href="#installation">Installation</a></li>
      <li><a href="#usage">Usage</a></li>
      <li><a href="#citation">Citation</a></li>
      <li><a href="#acknowledgments">Acknowledgments</a></li>
      <li><a href="#license">LICENSE</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Shixiang Wang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
