<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Read absolute copy number profile — read_copynumber • sigminer</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Read absolute copy number profile — read_copynumber" />

<meta property="og:description" content="Read absolute copy number profile for preparing CNV signature
analysis." />

<meta property="og:image" content="/logo.png" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sigminer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/sigminer.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ShixiangWang/sigminer">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Read absolute copy number profile</h1>
    <small class="dont-index">Source: <a href='https://github.com/ShixiangWang/sigminer/blob/master/R/read_series.R'><code>R/read_series.R</code></a></small>
    <div class="hidden name"><code>read_copynumber.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Read <strong>absolute</strong> copy number profile for preparing CNV signature
analysis.</p>
    
    </div>

    <pre class="usage"><span class='fu'>read_copynumber</span>(<span class='no'>input</span>, <span class='kw'>pattern</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>ignore_case</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>seg_cols</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"Chromosome"</span>, <span class='st'>"Start.bp"</span>, <span class='st'>"End.bp"</span>, <span class='st'>"modal_cn"</span>),
  <span class='kw'>samp_col</span> <span class='kw'>=</span> <span class='st'>"sample"</span>, <span class='kw'>use_all</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>min_segnum</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>genome_build</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"hg19"</span>, <span class='st'>"hg38"</span>), <span class='kw'>genome_measure</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"called"</span>, <span class='st'>"wg"</span>),
  <span class='kw'>clinical_data</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>complement</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>input</th>
      <td><p>a <code>data.frame</code> or a file or a directory contains copy number profile.</p></td>
    </tr>
    <tr>
      <th>pattern</th>
      <td><p>an optional regular expression used to select part of files if
<code>input</code> is a directory, more detail please see <a href='https://www.rdocumentation.org/packages/base/topics/list.files'>list.files</a> function.</p></td>
    </tr>
    <tr>
      <th>ignore_case</th>
      <td><p>logical. Should pattern-matching be case-insensitive?</p></td>
    </tr>
    <tr>
      <th>seg_cols</th>
      <td><p>four characters used to specify chromosome, start position,
end position and copy number value in <code>input</code>, respectively.
Default use names from ABSOLUTE calling result.</p></td>
    </tr>
    <tr>
      <th>samp_col</th>
      <td><p>a character used to specify the sample column name. If <code>input</code>
is a directory and cannot find <code>samp_col</code>, sample names will use file names
(set this parameter to <code>NULL</code> is recommended in this case).</p></td>
    </tr>
    <tr>
      <th>use_all</th>
      <td><p>default is <code>FALSE</code>. If <code>True</code>, use all columns from raw input.</p></td>
    </tr>
    <tr>
      <th>min_segnum</th>
      <td><p>minimal number of copy number segments within a sample.</p></td>
    </tr>
    <tr>
      <th>genome_build</th>
      <td><p>genome build version, should be 'hg19' or 'hg38'.</p></td>
    </tr>
    <tr>
      <th>genome_measure</th>
      <td><p>default is 'called', can be 'wg' or 'called'.
Set 'called' will use autosomo called segments size to compute total size for CNA burden calculation,
this option is useful for WES and target sequencing.
Set 'wg' will use autosome size from genome build, this option is useful for WGS, SNP etc..</p></td>
    </tr>
    <tr>
      <th>clinical_data</th>
      <td><p>a <code>data.frame</code> representing clinical data
associated with each sample in copy number profile.</p></td>
    </tr>
    <tr>
      <th>complement</th>
      <td><p>if <code>TRUE</code>, complement chromosome does not show in input data
with normal copy 2 and force <code>use_all</code> to <code>FALSE</code> (no matter what user input).</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>print extra messages.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>other parameters pass to <code><a href='https://www.rdocumentation.org/packages/data.table/topics/fread'>data.table::fread()</a></code></p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a <a href='CopyNumber-class.html'>CopyNumber</a> object</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other read genomic variation data function series: <code><a href='read_maf.html'>read_maf</a></code>,
  <code><a href='read_variation.html'>read_variation</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='co'># Load toy dataset of absolute copynumber profile</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/load'>load</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"toy_segTab.RData"</span>,
  <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"sigminer"</span>, <span class='kw'>mustWork</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
))
<span class='no'>cn</span> <span class='kw'>&lt;-</span> <span class='fu'>read_copynumber</span>(<span class='no'>segTabs</span>,
  <span class='kw'>seg_cols</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"chromosome"</span>, <span class='st'>"start"</span>, <span class='st'>"end"</span>, <span class='st'>"segVal"</span>),
  <span class='kw'>genome_build</span> <span class='kw'>=</span> <span class='st'>"hg19"</span>, <span class='kw'>complement</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Shixiang Wang <a href='mailto:w_shixiang@163.com'>w_shixiang@163.com</a></p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Shixiang Wang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

