<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Read MAF Files — read_maf • sigminer</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Read MAF Files — read_maf" />
<meta property="og:description" content="This function is a wrapper of maftools::read.maf.
Currently, only mutation records used in signature identification and
analysis.
You can also use maftools::read.maf to read the data." />
<meta property="og:image" content="/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sigminer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/sigminer.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ShixiangWang/sigminer/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Read MAF Files</h1>
    <small class="dont-index">Source: <a href='https://github.com/ShixiangWang/sigminer/blob/master/R/read_maf.R'><code>R/read_maf.R</code></a></small>
    <div class="hidden name"><code>read_maf.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function is a wrapper of <a href='https://rdrr.io/pkg/maftools/man/read.maf.html'>maftools::read.maf</a>.
Currently, only mutation records used in signature identification and
analysis.
You can also use <a href='https://rdrr.io/pkg/maftools/man/read.maf.html'>maftools::read.maf</a> to read the data.</p>
    </div>

    <pre class="usage"><span class='fu'>read_maf</span>(
  <span class='no'>maf</span>,
  <span class='kw'>clinicalData</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>removeDuplicatedVariants</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>useAll</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>gisticAllLesionsFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>gisticAmpGenesFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>gisticDelGenesFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>gisticScoresFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>cnLevel</span> <span class='kw'>=</span> <span class='st'>"all"</span>,
  <span class='kw'>cnTable</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>isTCGA</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>vc_nonSyn</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>maf</th>
      <td><p>tab delimited MAF file. File can also be gz compressed. Required. Alternatively, you can also provide already read MAF file as a dataframe.</p></td>
    </tr>
    <tr>
      <th>clinicalData</th>
      <td><p>Clinical data associated with each sample/Tumor_Sample_Barcode in MAF. Could be a text file or a data.frame. Default NULL.</p></td>
    </tr>
    <tr>
      <th>removeDuplicatedVariants</th>
      <td><p>removes repeated variants in a particuar sample, mapped to multiple transcripts of same Gene. See Description. Default TRUE.</p></td>
    </tr>
    <tr>
      <th>useAll</th>
      <td><p>logical. Whether to use all variants irrespective of values in Mutation_Status. Defaults to TRUE. If FALSE, only uses with values Somatic.</p></td>
    </tr>
    <tr>
      <th>gisticAllLesionsFile</th>
      <td><p>All Lesions file generated by gistic. e.g; all_lesions.conf_XX.txt, where XX is the confidence level. Default NULL.</p></td>
    </tr>
    <tr>
      <th>gisticAmpGenesFile</th>
      <td><p>Amplification Genes file generated by gistic. e.g; amp_genes.conf_XX.txt, where XX is the confidence level. Default NULL.</p></td>
    </tr>
    <tr>
      <th>gisticDelGenesFile</th>
      <td><p>Deletion Genes file generated by gistic. e.g; del_genes.conf_XX.txt, where XX is the confidence level. Default NULL.</p></td>
    </tr>
    <tr>
      <th>gisticScoresFile</th>
      <td><p>scores.gistic file generated by gistic. Default NULL</p></td>
    </tr>
    <tr>
      <th>cnLevel</th>
      <td><p>level of CN changes to use. Can be 'all', 'deep' or 'shallow'. Default uses all i.e, genes with both 'shallow' or 'deep' CN changes</p></td>
    </tr>
    <tr>
      <th>cnTable</th>
      <td><p>Custom copynumber data if gistic results are not available. Input file or a data.frame should contain three columns in aforementioned order with gene name, Sample name and copy number status (either 'Amp' or 'Del'). Default NULL.</p></td>
    </tr>
    <tr>
      <th>isTCGA</th>
      <td><p>Is input MAF file from TCGA source. If TRUE uses only first 12 characters from Tumor_Sample_Barcode.</p></td>
    </tr>
    <tr>
      <th>vc_nonSyn</th>
      <td><p>NULL. Provide manual list of variant classifications to be considered as non-synonymous. Rest will be considered as silent variants. Default uses Variant Classifications with High/Moderate variant consequences. http://asia.ensembl.org/Help/Glossary?id=535: "Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Site", "Translation_Start_Site","Nonsense_Mutation", "Nonstop_Mutation", "In_Frame_Del","In_Frame_Ins", "Missense_Mutation"</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>TRUE logical. Default to be talkative and prints summary.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='read_copynumber.html'>read_copynumber</a> for reading copy number data to <a href='CopyNumber-class.html'>CopyNumber</a> object.</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>laml.maf</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"tcga_laml.maf.gz"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"maftools"</span>, <span class='kw'>mustWork</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='no'>laml</span> <span class='kw'>&lt;-</span> <span class='fu'>read_maf</span>(<span class='kw'>maf</span> <span class='kw'>=</span> <span class='no'>laml.maf</span>)</div><div class='output co'>#&gt; -Reading
#&gt; -Validating
#&gt; -Silent variants: 475 
#&gt; -Summarizing
#&gt; -Processing clinical data
#&gt; --Missing clinical data
#&gt; -Finished in 0.280s elapsed (0.210s cpu) </div><div class='input'><span class='no'>laml</span></div><div class='output co'>#&gt; An object of class  MAF 
#&gt;                    ID          summary  Mean Median
#&gt;  1:        NCBI_Build               37    NA     NA
#&gt;  2:            Center genome.wustl.edu    NA     NA
#&gt;  3:           Samples              193    NA     NA
#&gt;  4:            nGenes             1241    NA     NA
#&gt;  5:   Frame_Shift_Del               52 0.271      0
#&gt;  6:   Frame_Shift_Ins               91 0.474      0
#&gt;  7:      In_Frame_Del               10 0.052      0
#&gt;  8:      In_Frame_Ins               42 0.219      0
#&gt;  9: Missense_Mutation             1342 6.990      7
#&gt; 10: Nonsense_Mutation              103 0.536      0
#&gt; 11:       Splice_Site               92 0.479      0
#&gt; 12:             total             1732 9.021      9</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://orcid.org/0000-0001-9855-7357'>Shixiang Wang</a>, <a href='https://orcid.org/0000-0002-7736-0077'>Xue-Song Liu</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.0.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


