<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Tally a Genomic Alteration Object — sig_tally • sigminer</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Tally a Genomic Alteration Object — sig_tally" />
<meta property="og:description" content="Tally a variation object like MAF, CopyNumber and return a matrix for NMF de-composition and more.
This is a generic function,
so it can be further extended to other mutation cases.
Please read details about how to set sex for identifying copy number signatures.
Please read https://osf.io/s93d5/ for the generation of SBS, DBS and ID (INDEL)
components. Of note, many options are designed for method &quot;M&quot; only, and they are highlighted
by bold fonts (you can ignore them if you don't use &quot;M&quot; method)." />
<meta property="og:image" content="/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sigminer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/sigminer.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ShixiangWang/sigminer/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Tally a Genomic Alteration Object</h1>
    <small class="dont-index">Source: <a href='https://github.com/ShixiangWang/sigminer/blob/master/R/sig_tally.R'><code>R/sig_tally.R</code></a></small>
    <div class="hidden name"><code>sig_tally.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Tally a variation object like <a href='MAF-class.html'>MAF</a>, <a href='CopyNumber-class.html'>CopyNumber</a> and return a matrix for NMF de-composition and more.
This is a generic function,
so it can be further extended to other mutation cases.
<strong>Please read details about how to set sex for identifying copy number signatures</strong>.
Please read <a href='https://osf.io/s93d5/'>https://osf.io/s93d5/</a> for the generation of SBS, DBS and ID (INDEL)
components. <strong>Of note, many options are designed for method "M" only, and they are highlighted
by bold fonts</strong> (you can ignore them if you don't use "M" method).</p>
    </div>

    <pre class="usage"><span class='fu'>sig_tally</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S3 method for CopyNumber</span>
<span class='fu'>sig_tally</span>(
  <span class='no'>object</span>,
  <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"Wang"</span>,
  <span class='kw'>ignore_chrs</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>indices</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>feature_setting</span> <span class='kw'>=</span> <span class='kw pkg'>sigminer</span><span class='kw ns'>::</span><span class='no'><a href='CN.features.html'>CN.features</a></span>,
  <span class='kw'>type</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"probability"</span>, <span class='st'>"count"</span>),
  <span class='kw'>reference_components</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>cores</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>seed</span> <span class='kw'>=</span> <span class='fl'>123456</span>,
  <span class='kw'>min_comp</span> <span class='kw'>=</span> <span class='fl'>2</span>,
  <span class='kw'>max_comp</span> <span class='kw'>=</span> <span class='fl'>15</span>,
  <span class='kw'>min_prior</span> <span class='kw'>=</span> <span class='fl'>0.001</span>,
  <span class='kw'>model_selection</span> <span class='kw'>=</span> <span class='st'>"BIC"</span>,
  <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0.1</span>,
  <span class='kw'>nrep</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>niter</span> <span class='kw'>=</span> <span class='fl'>1000</span>,
  <span class='kw'>keep_only_matrix</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='no'>...</span>
)

<span class='co'># S3 method for MAF</span>
<span class='fu'>sig_tally</span>(
  <span class='no'>object</span>,
  <span class='kw'>mode</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"SBS"</span>, <span class='st'>"DBS"</span>, <span class='st'>"ID"</span>, <span class='st'>"ALL"</span>),
  <span class='kw'>ref_genome</span> <span class='kw'>=</span> <span class='st'>"BSgenome.Hsapiens.UCSC.hg19"</span>,
  <span class='kw'>genome_build</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>add_trans_bias</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>ignore_chrs</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>use_syn</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>keep_only_matrix</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='no'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>a <a href='CopyNumber-class.html'>CopyNumber</a> object or <a href='MAF-class.html'>MAF</a> object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>custom setting for operating object. Detail see S3 method for
corresponding class (e.g. <code>CopyNumber</code>).</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>method for feature classfication, can be one of "Macintyre" ("M"),
"Wang" ("W").</p></td>
    </tr>
    <tr>
      <th>ignore_chrs</th>
      <td><p>Chromsomes to ignore from analysis. e.g. chrX and chrY.</p></td>
    </tr>
    <tr>
      <th>indices</th>
      <td><p>integer vector indicating segments to keep.</p></td>
    </tr>
    <tr>
      <th>feature_setting</th>
      <td><p>a <code>data.frame</code> used for classification.
<strong>Only used when method is "Wang" ("W")</strong>.
Default is <a href='CN.features.html'>CN.features</a>. Users can also set custom input with "feature",
"min" and "max" columns available. Valid features can be printed by
<code><a href='https://rdrr.io/r/base/unique.html'>unique(CN.features$feature)</a></code>.</p></td>
    </tr>
    <tr>
      <th>type</th>
      <td><p>one of "probability", "count". Default is "probability", return a matrix
with the sum of posterior probabilities for each components. If set to 'count',
return a matrix with event count assigned to each components. The result for
both types should be close. <strong>Only used when method is "Macintyre"</strong>.</p></td>
    </tr>
    <tr>
      <th>reference_components</th>
      <td><p>default is <code>FALSE</code>, calculate mixture components
from <a href='CopyNumber-class.html'>CopyNumber</a> object. <strong>Only used when method is "Macintyre"</strong>.</p></td>
    </tr>
    <tr>
      <th>cores</th>
      <td><p>number of computer cores to run this task.
You can use <code><a href='https://rdrr.io/pkg/future/man/availableCores.html'>future::availableCores()</a></code> function to check how
many cores you can use.</p></td>
    </tr>
    <tr>
      <th>seed</th>
      <td><p>seed number. <strong>Only used when method is "Macintyre"</strong>.</p></td>
    </tr>
    <tr>
      <th>min_comp</th>
      <td><p>minimal number of components to fit, default is 2.
Can also be a vector with length 6, which apply to each feature.
<strong>Only used when method is "Macintyre"</strong>.</p></td>
    </tr>
    <tr>
      <th>max_comp</th>
      <td><p>maximal number of components to fit, default is 15.
Can also be a vector with length 6, which apply to each feature.
<strong>Only used when method is "Macintyre"</strong>.</p></td>
    </tr>
    <tr>
      <th>min_prior</th>
      <td><p>the minimum relative size of components, default is 0.001.
Details about custom setting please refer to <strong>flexmix</strong> package.
<strong>Only used when method is "Macintyre"</strong>.</p></td>
    </tr>
    <tr>
      <th>model_selection</th>
      <td><p>model selection strategy, default is 'BIC'.
Details about custom setting please refer to <strong>flexmix</strong> package.
<strong>Only used when method is "Macintyre"</strong>.</p></td>
    </tr>
    <tr>
      <th>threshold</th>
      <td><p>default is <code>0.1</code>. Sometimes, the result components
include adjacent distributions with similar mu
(two and more distribution are very close), we use this threshold
to obtain a more meaningful fit with less components.
<strong>Only used when method is "Macintyre"</strong>.</p></td>
    </tr>
    <tr>
      <th>nrep</th>
      <td><p>number of run times for each value of component,
keep only the solution with maximum likelihood.
<strong>Only used when method is "Macintyre"</strong>.</p></td>
    </tr>
    <tr>
      <th>niter</th>
      <td><p>the maximum number of iterations.
<strong>Only used when method is "Macintyre"</strong>.</p></td>
    </tr>
    <tr>
      <th>keep_only_matrix</th>
      <td><p>if <code>TRUE</code>, keep only matrix for signature extraction.
For a <code>MAF</code> object, this will just return the most useful matrix.</p></td>
    </tr>
    <tr>
      <th>mode</th>
      <td><p>type of mutation matrix to extract, can be one of 'SBS', 'DBS' and 'ID'.</p></td>
    </tr>
    <tr>
      <th>ref_genome</th>
      <td><p>'BSgenome.Hsapiens.UCSC.hg19', 'BSgenome.Hsapiens.UCSC.hg38' and
'BSgenome.Mmusculus.UCSC.mm10' etc.</p></td>
    </tr>
    <tr>
      <th>genome_build</th>
      <td><p>genome build 'hg19', 'hg38' or "mm10", if not set, guess it by <code>ref_genome</code>.</p></td>
    </tr>
    <tr>
      <th>add_trans_bias</th>
      <td><p>if <code>TRUE</code>, consider transcriptional bias categories.
'T:' for Transcribed (the variant is on the transcribed strand);
'U:' for Un-transcribed (the variant is on the untranscribed strand);
'B:' for Bi-directional (the variant is on both strand and is transcribed either way);
'N:' for Non-transcribed (the variant is in a non-coding region and is untranslated);
'Q:' for Questionable.
<strong>NOTE</strong>: the result counts of 'B' and 'N' labels are a little different from
SigProfilerMatrixGenerator, the reason is unknown (may be caused by annotation file).</p></td>
    </tr>
    <tr>
      <th>use_syn</th>
      <td><p>Logical. Whether to include synonymous variants in analysis. Defaults to TRUE</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a <code>list</code> contains a <code>matrix</code> used for NMF de-composition.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>For identifying copy number signatures, we have to derive copy number
features firstly. Due to the difference of copy number values in sex chromosomes
between male and female, we have to do an extra step <strong>if we don't want to
ignore them</strong>.</p>
<p>I create two options to control this, the default values are shown as
the following, you can use the same way to set (per R session).</p>
<p><code><a href='https://rdrr.io/r/base/options.html'>options(sigminer.sex = "female", sigminer.copynumber.max = NA_integer_)</a></code></p><ul>
<li><p>If your cohort are all females, you can totally ignore this.</p></li>
<li><p>If your cohort are all males, set <code>sigminer.sex</code> to 'male' and
<code>sigminer.copynumber.max</code> to a proper value (the best is consistent
with <a href='read_copynumber.html'>read_copynumber</a>).</p></li>
<li><p>If your cohort contains both males and females, set <code>sigminer.sex</code>
as a <code>data.frame</code> with two columns "sample" and "sex". And
set <code>sigminer.copynumber.max</code> to a proper value (the best is consistent
with <a href='read_copynumber.html'>read_copynumber</a>).</p></li>
</ul>

    <h2 class="hasAnchor" id="methods-by-class-"><a class="anchor" href="#methods-by-class-"></a>Methods (by class)</h2>

    
<ul>
<li><p><code>CopyNumber</code>: Returns copy number features, components and component-by-sample matrix</p></li>
<li><p><code>MAF</code>: Returns SBS mutation sample-by-component matrix and APOBEC enrichment</p></li>
</ul>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Macintyre, Geoff, et al. "Copy number signatures and mutational
processes in ovarian carcinoma." Nature genetics 50.9 (2018): 1262.</p>
<p>Wang, Shixiang, et al. "Copy number signature analyses in prostate cancer reveal
distinct etiologies and clinical outcomes." medRxiv (2020).</p>
<p>Mayakonda, Anand, et al. "Maftools: efficient and comprehensive analysis of somatic variants in cancer." Genome research 28.11 (2018): 1747-1756.</p>
<p>Roberts SA, Lawrence MS, Klimczak LJ, et al. An APOBEC Cytidine Deaminase Mutagenesis Pattern is Widespread in Human Cancers. Nature genetics. 2013;45(9):970-976. doi:10.1038/ng.2702.</p>
<p>Bergstrom EN, Huang MN, Mahto U, Barnes M, Stratton MR, Rozen SG, Alexandrov LB: SigProfilerMatrixGenerator: a tool for visualizing and exploring patterns of small mutational events. BMC Genomics 2019, 20:685 https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-019-6041-2</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='sig_estimate.html'>sig_estimate</a> for estimating signature number for <a href='sig_extract.html'>sig_extract</a>,
<a href='sig_auto_extract.html'>sig_auto_extract</a> for extracting signatures using automatic relevance determination technique.</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Load copy number object</span>
<span class='fu'><a href='https://rdrr.io/r/base/load.html'>load</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"toy_copynumber.RData"</span>,
  <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"sigminer"</span>, <span class='kw'>mustWork</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
))
<span class='co'># \donttest{</span>
<span class='co'># Use method designed by Wang, Shixiang et al.</span>
<span class='no'>cn_tally_W</span> <span class='kw'>&lt;-</span> <span class='fu'>sig_tally</span>(<span class='no'>cn</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"W"</span>)</div><div class='output co'>#&gt; <span class='message'><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:34]: </span><span style='color: #0000BB;font-weight: bold;'>Started.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:34]: </span><span style='color: #0000BB;font-weight: bold;'>Step: getting copy number features.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:34]: </span><span style='color: #0000BB;font-weight: bold;'>Getting breakpoint count per 10 Mb...</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:34]: </span><span style='color: #0000BB;font-weight: bold;'>Getting breakpoint count per chromosome arm...</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:34]: </span><span style='color: #0000BB;font-weight: bold;'>Getting copy number...</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:34]: </span><span style='color: #0000BB;font-weight: bold;'>Getting change-point copy number change...</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:35]: </span><span style='color: #0000BB;font-weight: bold;'>Getting length of chains of oscillating copy number...</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:35]: </span><span style='color: #0000BB;font-weight: bold;'>Getting (log10 based) segment size...</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:35]: </span><span style='color: #0000BB;font-weight: bold;'>Getting the minimal number of chromosome with 50% CNV...</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:35]: </span><span style='color: #0000BB;font-weight: bold;'>Getting burden of chromosome...</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:35]: </span><span style='color: #00BB00;font-weight: bold;'>Gotten.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:35]: </span><span style='color: #0000BB;font-weight: bold;'>Step: generating copy number components.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:35]: </span><span style='color: #000000;background-color: #BBBBBB;font-weight: bold;'>`feature_setting`</span><span style='font-weight: bold;'> checked.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:35]: </span><span style='color: #0000BB;font-weight: bold;'>Step: counting components.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:36]: </span><span style='color: #00BB00;font-weight: bold;'>Counted.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:36]: </span><span style='color: #0000BB;font-weight: bold;'>Step: generating components by sample matrix.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:36]: </span><span style='color: #00BB00;font-weight: bold;'>Matrix generated.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:36]: </span><span style='color: #0000BB;font-weight: bold;'>1.678 secs elapsed.</span><span></span></div><div class='input'><span class='co'># Use method designed by Macintyre et al.</span>
<span class='no'>cn_tally_M</span> <span class='kw'>&lt;-</span> <span class='fu'>sig_tally</span>(<span class='no'>cn</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"M"</span>)</div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:36]: </span><span style='color: #0000BB;font-weight: bold;'>Started.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:36]: </span><span style='color: #0000BB;font-weight: bold;'>Step: getting copy number features.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:36]: </span><span style='color: #0000BB;font-weight: bold;'>Getting segment size...</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:36]: </span><span style='color: #0000BB;font-weight: bold;'>Getting breakpoint count per 10 Mb...</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:36]: </span><span style='color: #0000BB;font-weight: bold;'>Getting length of chains of oscillating copy number...</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:36]: </span><span style='color: #0000BB;font-weight: bold;'>Getting breakpoint count per chromosome arm...</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:36]: </span><span style='color: #0000BB;font-weight: bold;'>Getting change-point copy number change...</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:36]: </span><span style='color: #0000BB;font-weight: bold;'>Getting copy number...</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:37]: </span><span style='color: #00BB00;font-weight: bold;'>Gotten.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:37]: </span><span style='color: #0000BB;font-weight: bold;'>Step: fitting copy number components.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:37]: </span><span style='color: #0000BB;font-weight: bold;'>Fitting feature: segsize.</span><span></span></div><div class='output co'>#&gt; <span class='message'>载入需要的程辑包：globals</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>载入程辑包：‘globals’</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:R.utils’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    cleanup</span></div><div class='output co'>#&gt; <span class='message'>载入需要的程辑包：future</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>载入程辑包：‘future’</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:rtracklayer’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    values</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:GenomicRanges’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    values</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:IRanges’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    values</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:S4Vectors’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    values</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:NMF’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    run</span></div><div class='output co'>#&gt; <span class='message'>载入需要的程辑包：lattice</span></div><div class='output co'>#&gt; 2 : *
#&gt; 3 : *
#&gt; 4 : *
#&gt; 5 : *
#&gt; 6 : *
#&gt; 7 : *
#&gt; 8 : *
#&gt; 9 : *
#&gt; 10 : *
#&gt; 11 : *
#&gt; 12 : *
#&gt; 13 : *
#&gt; 14 : *
#&gt; 15 : *</div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:45]: </span><span style='color: #00BB00;font-weight: bold;'>Select 10 according to BIC.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:45]: </span><span style='color: #00BB00;font-weight: bold;'>Finally, select 10 after passing threshold 0.1.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:45]: </span><span style='color: #0000BB;font-weight: bold;'>Fitting feature: bp10MB.</span><span></span></div><div class='output co'>#&gt; 2 : *
#&gt; 3 : *
#&gt; 4 : *
#&gt; 5 : *
#&gt; 6 : *
#&gt; 7 : *
#&gt; 8 : *
#&gt; 9 : *
#&gt; 10 : *
#&gt; 11 : *
#&gt; 12 : *
#&gt; 13 : *
#&gt; 14 : *
#&gt; 15 : *</div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:49]: </span><span style='color: #00BB00;font-weight: bold;'>Select 2 according to BIC.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:49]: </span><span style='color: #00BB00;font-weight: bold;'>Finally, select 2 after passing threshold 0.1.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:49]: </span><span style='color: #0000BB;font-weight: bold;'>Fitting feature: osCN.</span><span></span></div><div class='output co'>#&gt; 2 : *
#&gt; 3 : *
#&gt; 4 : *
#&gt; 5 : *
#&gt; 6 : *
#&gt; 7 : *
#&gt; 8 : *
#&gt; 9 : *
#&gt; 10 : *
#&gt; 11 : *
#&gt; 12 : *
#&gt; 13 : *
#&gt; 14 : *
#&gt; 15 : *</div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:51]: </span><span style='color: #00BB00;font-weight: bold;'>Select 2 according to BIC.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:51]: </span><span style='color: #00BB00;font-weight: bold;'>Finally, select 2 after passing threshold 0.1.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:51]: </span><span style='color: #0000BB;font-weight: bold;'>Fitting feature: bpchrarm.</span><span></span></div><div class='output co'>#&gt; 2 : *
#&gt; 3 : *
#&gt; 4 : *
#&gt; 5 : *
#&gt; 6 : *
#&gt; 7 : *
#&gt; 8 : *
#&gt; 9 : *
#&gt; 10 : *
#&gt; 11 : *
#&gt; 12 : *
#&gt; 13 : *
#&gt; 14 : *
#&gt; 15 : *</div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:54]: </span><span style='color: #00BB00;font-weight: bold;'>Select 2 according to BIC.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:54]: </span><span style='color: #00BB00;font-weight: bold;'>Finally, select 2 after passing threshold 0.1.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:54]: </span><span style='color: #0000BB;font-weight: bold;'>Fitting feature: changepoint.</span><span></span></div><div class='output co'>#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     25 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     19 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     14 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     14 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     13 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     15 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     12 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     14 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     15 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     20 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     15 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     14 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     14 Log-likelihood: NaN
#&gt; 2 : *
#&gt; 3 : *
#&gt; 4 : *
#&gt; 5 : *
#&gt; 6 : *
#&gt; 7 : *
#&gt; 8 : *
#&gt; 9 : *
#&gt; 10 : *
#&gt; 11 : *
#&gt; 12 : *
#&gt; 13 : *
#&gt; 14 : *
#&gt; 15 : *</div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:55]: </span><span style='color: #0000BB;font-weight: bold;'>Fitting feature: copynumber.</span><span></span></div><div class='output co'>#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     12 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     10 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;      9 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     12 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     11 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     10 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     10 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     12 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     10 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     10 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     11 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     11 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     10 Log-likelihood: NaN
#&gt; 2 : *
#&gt; 3 : *
#&gt; 4 : *
#&gt; 5 : *
#&gt; 6 : *
#&gt; 7 : *
#&gt; 8 : *
#&gt; 9 : *
#&gt; 10 : *
#&gt; 11 : *
#&gt; 12 : *
#&gt; 13 : *
#&gt; 14 : *
#&gt; 15 : *</div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:55]: </span><span style='color: #00BB00;font-weight: bold;'>Components fitted.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:55]: </span><span style='color: #0000BB;font-weight: bold;'>Step: calculating the sum of posterior probabilities.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:56]: </span><span style='color: #0000BB;font-weight: bold;'>Calculating the sum of posterior probabilities for bp10MB.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:56]: </span><span style='color: #0000BB;font-weight: bold;'>Calculating the sum of posterior probabilities for copynumber.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:56]: </span><span style='color: #0000BB;font-weight: bold;'>Calculating the sum of posterior probabilities for changepoint.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:56]: </span><span style='color: #0000BB;font-weight: bold;'>Calculating the sum of posterior probabilities for bpchrarm.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:56]: </span><span style='color: #0000BB;font-weight: bold;'>Calculating the sum of posterior probabilities for osCN.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:56]: </span><span style='color: #0000BB;font-weight: bold;'>Calculating the sum of posterior probabilities for segsize.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:56]: </span><span style='color: #00BB00;font-weight: bold;'>Matrix generated.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:56]: </span><span style='color: #0000BB;font-weight: bold;'>20.399 secs elapsed.</span><span></span></div><div class='input'><span class='co'># }</span>
<span class='co'># \donttest{</span>
<span class='co'># Prepare SBS signature analysis</span>
<span class='no'>laml.maf</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"tcga_laml.maf.gz"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"maftools"</span>)
<span class='no'>laml</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='read_maf.html'>read_maf</a></span>(<span class='kw'>maf</span> <span class='kw'>=</span> <span class='no'>laml.maf</span>)</div><div class='output co'>#&gt; -Reading
#&gt; -Validating
#&gt; -Silent variants: 475 
#&gt; -Summarizing
#&gt; -Processing clinical data
#&gt; --Missing clinical data
#&gt; -Finished in 0.217s elapsed (0.209s cpu) </div><div class='input'><span class='kw'>if</span> (<span class='fu'><a href='https://rdrr.io/r/base/library.html'>require</a></span>(<span class='st'>"BSgenome.Hsapiens.UCSC.hg19"</span>)) {
  <span class='no'>mt_tally</span> <span class='kw'>&lt;-</span> <span class='fu'>sig_tally</span>(
    <span class='no'>laml</span>,
    <span class='kw'>ref_genome</span> <span class='kw'>=</span> <span class='st'>"BSgenome.Hsapiens.UCSC.hg19"</span>,
    <span class='kw'>use_syn</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
  )
  <span class='no'>mt_tally</span>$<span class='no'>nmf_matrix</span>[<span class='fl'>1</span>:<span class='fl'>5</span>, <span class='fl'>1</span>:<span class='fl'>5</span>]

  <span class='co'>## Use strand bias categories</span>
  <span class='no'>mt_tally</span> <span class='kw'>&lt;-</span> <span class='fu'>sig_tally</span>(
    <span class='no'>laml</span>,
    <span class='kw'>ref_genome</span> <span class='kw'>=</span> <span class='st'>"BSgenome.Hsapiens.UCSC.hg19"</span>,
    <span class='kw'>use_syn</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>add_trans_bias</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
  )
  <span class='co'>## Test it by enrichment analysis</span>
  <span class='fu'><a href='enrich_component_strand_bias.html'>enrich_component_strand_bias</a></span>(<span class='no'>mt_tally</span>$<span class='no'>nmf_matrix</span>)
  <span class='fu'><a href='enrich_component_strand_bias.html'>enrich_component_strand_bias</a></span>(<span class='no'>mt_tally</span>$<span class='no'>all_matrices</span>$<span class='no'>SBS_24</span>)
} <span class='kw'>else</span> {
  <span class='fu'><a href='https://rdrr.io/r/base/message.html'>message</a></span>(<span class='st'>"Please install package 'BSgenome.Hsapiens.UCSC.hg19' firstly!"</span>)
}</div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:56]: </span><span style='color: #0000BB;font-weight: bold;'>Started.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:56]: </span><span style='color: #00BB00;font-weight: bold;'>Reference genome loaded.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:56]: </span><span style='color: #00BB00;font-weight: bold;'>Variants from MAF object queried.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:56]: </span><span style='color: #00BB00;font-weight: bold;'>Chromosome names checked.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:56]: </span><span style='color: #00BB00;font-weight: bold;'>Sex chromosomes properly handled.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:56]: </span><span style='color: #00BB00;font-weight: bold;'>Variant start and end position checked.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:56]: </span><span style='color: #00BB00;font-weight: bold;'>Variant data for matrix generation preprocessed.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:56]: </span><span style='color: #0000BB;font-weight: bold;'>SBS matrix generation - start.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:56]: </span><span style='color: #0000BB;font-weight: bold;'>Extracting 5' and 3' adjacent bases.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:57]: </span><span style='color: #0000BB;font-weight: bold;'>Extracting +/- 20bp around mutated bases for background C&gt;T estimation.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #0000BB;font-weight: bold;'>Estimating APOBEC enrichment scores.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #0000BB;font-weight: bold;'>Performing one-way Fisher's test for APOBEC enrichment.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #00BB00;font-weight: bold;'>APOBEC related mutations are enriched in 3.297% of samples (APOBEC enrichment score &gt; 2; 6 of 182 samples)</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #0000BB;font-weight: bold;'>Creating SBS sample-by-component matrices.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #00BB00;font-weight: bold;'>SBS-6 matrix created.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #00BB00;font-weight: bold;'>SBS-96 matrix created.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #00BB00;font-weight: bold;'>SBS-1536 matrix created.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #0000BB;font-weight: bold;'>Return SBS-96 as major matrix.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #00BB00;font-weight: bold;'>Done.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #0000BB;font-weight: bold;'>1.65 secs elapsed.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #0000BB;font-weight: bold;'>Started.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #00BB00;font-weight: bold;'>Reference genome loaded.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #00BB00;font-weight: bold;'>Variants from MAF object queried.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #00BB00;font-weight: bold;'>Chromosome names checked.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #00BB00;font-weight: bold;'>Sex chromosomes properly handled.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #00BB00;font-weight: bold;'>Variant start and end position checked.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #00BB00;font-weight: bold;'>Variant data for matrix generation preprocessed.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #0000BB;font-weight: bold;'>SBS matrix generation - start.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #0000BB;font-weight: bold;'>Extracting 5' and 3' adjacent bases.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:58]: </span><span style='color: #0000BB;font-weight: bold;'>Extracting +/- 20bp around mutated bases for background C&gt;T estimation.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:59]: </span><span style='color: #0000BB;font-weight: bold;'>Estimating APOBEC enrichment scores.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:59]: </span><span style='color: #0000BB;font-weight: bold;'>Performing one-way Fisher's test for APOBEC enrichment.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:59]: </span><span style='color: #00BB00;font-weight: bold;'>APOBEC related mutations are enriched in 3.297% of samples (APOBEC enrichment score &gt; 2; 6 of 182 samples)</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:59]: </span><span style='color: #0000BB;font-weight: bold;'>Creating SBS sample-by-component matrices.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:59]: </span><span style='color: #00BB00;font-weight: bold;'>SBS-6 matrix created.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:59]: </span><span style='color: #00BB00;font-weight: bold;'>SBS-96 matrix created.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:59]: </span><span style='color: #00BB00;font-weight: bold;'>SBS-1536 matrix created.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:59]: </span><span style='color: #00BB00;font-weight: bold;'>SBS-24 (6x4) matrix created.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:59]: </span><span style='color: #00BB00;font-weight: bold;'>SBS-384 (96x4) matrix created.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:59]: </span><span style='color: #00BB00;font-weight: bold;'>SBS-6144 (1536x4) matrix created.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:59]: </span><span style='color: #0000BB;font-weight: bold;'>Return SBS-384 as major matrix.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BB00;'>✓</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:59]: </span><span style='color: #00BB00;font-weight: bold;'>Done.</span><span></span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='font-weight: bold;'>[2020-07-07 16:50:59]: </span><span style='color: #0000BB;font-weight: bold;'>1.381 secs elapsed.</span><span></span></div><div class='output co'>#&gt;             sample component N_T N_U Total_T Total_U Enrichment    p_value
#&gt;    1: TCGA-AB-2838       C&gt;T   3   7      12       8  0.4285714 0.01976661
#&gt;    2: TCGA-AB-3002       C&gt;T   5  11      13      14  0.4545455 0.05423790
#&gt;    3: TCGA-AB-2950       C&gt;T   4   2       4       6  2.0000000 0.07619048
#&gt;    4: TCGA-AB-2878       C&gt;T   3   4       8       4  0.7500000 0.08080808
#&gt;    5: TCGA-AB-2863       T&gt;C   3   0       8       9        Inf 0.08235294
#&gt;   ---                                                                     
#&gt; 1154: TCGA-AB-3012       T&gt;A   0   0       5       2        NaN 1.00000000
#&gt; 1155: TCGA-AB-3012       C&gt;G   0   0       5       2        NaN 1.00000000
#&gt; 1156: TCGA-AB-2876       C&gt;T   3   4       5       6  0.7500000 1.00000000
#&gt; 1157: TCGA-AB-2920       C&gt;T   6   2       7       2  3.0000000 1.00000000
#&gt; 1158: TCGA-AB-2920       C&gt;G   1   0       7       2        Inf 1.00000000
#&gt;             fdr
#&gt;    1: 0.1185997
#&gt;    2: 0.3254274
#&gt;    3: 0.4571429
#&gt;    4: 0.4848485
#&gt;    5: 0.4941176
#&gt;   ---          
#&gt; 1154: 1.0000000
#&gt; 1155: 1.0000000
#&gt; 1156: 1.0000000
#&gt; 1157: 1.0000000
#&gt; 1158: 1.0000000</div><div class='input'># }
</div></span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://orcid.org/0000-0001-9855-7357'>Shixiang Wang</a>, <a href='https://orcid.org/0000-0002-7736-0077'>Xue-Song Liu</a>, Tao Wu, Ziyu Tao.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


