<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Tally a Variation Object — sig_tally • sigminer</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Tally a Variation Object — sig_tally" />
<meta property="og:description" content="Tally a variation object like MAF and return a matrix for NMF de-composition and more.
This is a generic function,
so it can be further extended to other mutation cases. Please read details
about how to set sex for identifying copy number signatures." />
<meta property="og:image" content="/logo.png" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sigminer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/sigminer.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ShixiangWang/sigminer">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Tally a Variation Object</h1>
    <small class="dont-index">Source: <a href='https://github.com/ShixiangWang/sigminer/blob/master/R/sig_tally.R'><code>R/sig_tally.R</code></a></small>
    <div class="hidden name"><code>sig_tally.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Tally a variation object like <a href='MAF-class.html'>MAF</a> and return a matrix for NMF de-composition and more.
This is a generic function,
so it can be further extended to other mutation cases. Please read details
about how to set sex for identifying copy number signatures.</p>
    </div>

    <pre class="usage"><span class='fu'>sig_tally</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S3 method for CopyNumber</span>
<span class='fu'>sig_tally</span>(
  <span class='no'>object</span>,
  <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"Macintyre"</span>,
  <span class='kw'>ignore_chrs</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>feature_setting</span> <span class='kw'>=</span> <span class='kw pkg'>sigminer</span><span class='kw ns'>::</span><span class='no'><a href='https://rdrr.io/pkg/sigminer/man/CN.features.html'>CN.features</a></span>,
  <span class='kw'>type</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"probability"</span>, <span class='st'>"count"</span>),
  <span class='kw'>reference_components</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>cores</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>seed</span> <span class='kw'>=</span> <span class='fl'>123456</span>,
  <span class='kw'>min_comp</span> <span class='kw'>=</span> <span class='fl'>2</span>,
  <span class='kw'>max_comp</span> <span class='kw'>=</span> <span class='fl'>15</span>,
  <span class='kw'>min_prior</span> <span class='kw'>=</span> <span class='fl'>0.001</span>,
  <span class='kw'>model_selection</span> <span class='kw'>=</span> <span class='st'>"BIC"</span>,
  <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0.1</span>,
  <span class='kw'>nrep</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>niter</span> <span class='kw'>=</span> <span class='fl'>1000</span>,
  <span class='kw'>keep_only_matrix</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='no'>...</span>
)

<span class='co'># S3 method for MAF</span>
<span class='fu'>sig_tally</span>(
  <span class='no'>object</span>,
  <span class='kw'>ref_genome</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>prefix</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>add</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>ignore_chrs</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>use_syn</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>keep_only_matrix</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='no'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>a <a href='CopyNumber-class.html'>CopyNumber</a> object or <a href='MAF-class.html'>MAF</a> object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>custom setting for operating object. Detail see S3 method for
corresponding class (e.g. <code>CopyNumber</code>).</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>method for feature classfication, can be one of "Macintyre" ("M") and
"Wang" ("W").</p></td>
    </tr>
    <tr>
      <th>ignore_chrs</th>
      <td><p>Chromsomes to ignore from analysis. e.g. chrX and chrY.</p></td>
    </tr>
    <tr>
      <th>feature_setting</th>
      <td><p>a <code>data.frame</code> used for classification.
Only used when method is "Wang" ("W").
Default is <a href='CN.features.html'>CN.features</a>. Users can also set custom input with "feature",
"min" and "max" columns available. Valid features can be printed by
<code><a href='https://rdrr.io/r/base/unique.html'>unique(CN.features$feature)</a></code>.</p></td>
    </tr>
    <tr>
      <th>type</th>
      <td><p>one of "probability", "count". Default is "probability", return a matrix
with the sum of posterior probabilities for each components. If set to 'count',
return a matrix with event count assigned to each components. The result for
both types should be close. Only used when method is "Macintyre".</p></td>
    </tr>
    <tr>
      <th>reference_components</th>
      <td><p>default is <code>FALSE</code>, calculate mixture components
from <a href='CopyNumber-class.html'>CopyNumber</a> object. Only used when method is "Macintyre".</p></td>
    </tr>
    <tr>
      <th>cores</th>
      <td><p>number of compute cores to run this task.
You can use <code><a href='https://rdrr.io/pkg/future/man/availableCores.html'>future::availableCores()</a></code> function to check how
many cores you can use.</p></td>
    </tr>
    <tr>
      <th>seed</th>
      <td><p>seed number. Only used when method is "Macintyre".</p></td>
    </tr>
    <tr>
      <th>min_comp</th>
      <td><p>minimal number of components to fit, default is 2.
Can also be a vector with length 6, which apply to each feature.
Only used when method is "Macintyre".</p></td>
    </tr>
    <tr>
      <th>max_comp</th>
      <td><p>maximal number of components to fit, default is 15.
Can also be a vector with length 6, which apply to each feature.
Only used when method is "Macintyre".</p></td>
    </tr>
    <tr>
      <th>min_prior</th>
      <td><p>the minimum relative size of components, default is 0.001.
Details about custom setting please refer to <strong>flexmix</strong> package.
Only used when method is "Macintyre".</p></td>
    </tr>
    <tr>
      <th>model_selection</th>
      <td><p>model selection strategy, default is 'BIC'.
Details about custom setting please refer to <strong>flexmix</strong> package.
Only used when method is "Macintyre".</p></td>
    </tr>
    <tr>
      <th>threshold</th>
      <td><p>default is <code>0.1</code>. Sometimes, the result components
include adjacent distributions with similar mu
(two and more distribution are very close), we use this threshold
to obtain a more meaningful fit with less components.
Only used when method is "Macintyre".</p></td>
    </tr>
    <tr>
      <th>nrep</th>
      <td><p>number of run times for each value of component,
keep only the solution with maximum likelihood.
Only used when method is "Macintyre".</p></td>
    </tr>
    <tr>
      <th>niter</th>
      <td><p>the maximum number of iterations.
Only used when method is "Macintyre".</p></td>
    </tr>
    <tr>
      <th>keep_only_matrix</th>
      <td><p>if <code>TRUE</code>, keep only matrix for signature extraction.</p></td>
    </tr>
    <tr>
      <th>ref_genome</th>
      <td><p>BSgenome object or name of the installed BSgenome package. Example: BSgenome.Hsapiens.UCSC.hg19
Default NULL, tries to auto-detect from installed genomes.</p></td>
    </tr>
    <tr>
      <th>prefix</th>
      <td><p>Prefix to add or remove from contig names in MAF file.</p></td>
    </tr>
    <tr>
      <th>add</th>
      <td><p>If prefix is used, default is to add prefix to contig names in MAF file. If false prefix will be removed from contig names.</p></td>
    </tr>
    <tr>
      <th>use_syn</th>
      <td><p>Logical. Whether to include synonymous variants in analysis. Defaults to TRUE</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a <code>list</code> contains a <code>matrix</code> used for NMF de-composition.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>For identifying copy number signatures, we have to derive copy number
features firstly. Due to the difference of copy number values in sex chromosomes
between male and female, we have to do an extra step <strong>if we don't want to
ignore them</strong>.</p>
<p>I create two options to control this, the default values are shown as
the following, you can use the same way to set (per R session).</p>
<p><code><a href='https://rdrr.io/r/base/options.html'>options(sigminer.sex = "female", sigminer.copynumber.max = NA_integer_)</a></code></p><ul>
<li><p>If your cohort are all females, you can totally ignore this.</p></li>
<li><p>If your cohort are all males, set <code>sigminer.sex</code> to 'male' and
<code>sigminer.copynumber.max</code> to a proper value (the best is consistent
with <a href='read_copynumber.html'>read_copynumber</a>).</p></li>
<li><p>If your cohort contains bother males and females, set <code>sigminer.sex</code>
as a <code>data.frame</code> with two columns "sample" and "sex". And
set <code>sigminer.copynumber.max</code> to a proper value (the best is consistent
with <a href='read_copynumber.html'>read_copynumber</a>).</p></li>
</ul>

    <h2 class="hasAnchor" id="methods-by-class-"><a class="anchor" href="#methods-by-class-"></a>Methods (by class)</h2>

    
<ul>
<li><p><code>CopyNumber</code>: Returns copy number features, components and component-by-sample matrix</p></li>
<li><p><code>MAF</code>: Returns SBS mutation component-by-sample matrix and APOBEC enrichment</p></li>
</ul>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Macintyre, Geoff, et al. "Copy number signatures and mutational
processes in ovarian carcinoma." Nature genetics 50.9 (2018): 1262.</p>
<p>Mayakonda, Anand, et al. "Maftools: efficient and comprehensive analysis of somatic variants in cancer." Genome research 28.11 (2018): 1747-1756.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='sig_estimate.html'>sig_estimate</a> for estimating signature number for <a href='sig_extract.html'>sig_extract</a>,
<a href='sig_auto_extract.html'>sig_auto_extract</a> for extracting signatures using automatic relevance determination technique.</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Load copy number object</span>
<span class='fu'><a href='https://rdrr.io/r/base/load.html'>load</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"toy_copynumber.RData"</span>,
  <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"sigminer"</span>, <span class='kw'>mustWork</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
))
<span class='co'># \donttest{</span>
<span class='co'># Use method designed by Macintyre et al.</span>
<span class='no'>cn_tally_M</span> <span class='kw'>&lt;-</span> <span class='fu'>sig_tally</span>(<span class='no'>cn</span>)</div><div class='output co'>#&gt; <span class='message'>=&gt; Step: getting copy number features</span></div><div class='output co'>#&gt; <span class='message'>Getting segment size...</span></div><div class='output co'>#&gt; <span class='message'>Getting breakpoint count per 10 Mb...</span></div><div class='output co'>#&gt; <span class='message'>Getting length of chains of oscillating copy number...</span></div><div class='output co'>#&gt; <span class='message'>Getting breakpoint count per chromosome arm...</span></div><div class='output co'>#&gt; <span class='message'>Getting change-point copy number change...</span></div><div class='output co'>#&gt; <span class='message'>Getting copy number...</span></div><div class='output co'>#&gt; <span class='message'>=&gt; Step: fitting copy number components</span></div><div class='output co'>#&gt; <span class='message'>Fitting feature: segsize</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: globals</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: future</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: 'future'</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from 'package:NMF':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    run</span></div><div class='output co'>#&gt; 2 : *
#&gt; 3 : *
#&gt; 4 : *
#&gt; 5 : *
#&gt; 6 : *
#&gt; 7 : *
#&gt; 8 : *
#&gt; 9 : *
#&gt; 10 : *
#&gt; 11 : *
#&gt; 12 : *
#&gt; 13 : *
#&gt; 14 : *
#&gt; 15 : *</div><div class='output co'>#&gt; <span class='message'>Select 10 according to BIC</span></div><div class='output co'>#&gt; <span class='message'>Finally, select 10 after passing threshold 0.1</span></div><div class='output co'>#&gt; <span class='message'>===================</span></div><div class='output co'>#&gt; <span class='message'>Fitting feature: bp10MB</span></div><div class='output co'>#&gt; 2 : *
#&gt; 3 : *
#&gt; 4 : *
#&gt; 5 : *
#&gt; 6 : *
#&gt; 7 : *
#&gt; 8 : *
#&gt; 9 : *
#&gt; 10 : *
#&gt; 11 : *
#&gt; 12 : *
#&gt; 13 : *
#&gt; 14 : *
#&gt; 15 : *</div><div class='output co'>#&gt; <span class='message'>Select 2 according to BIC</span></div><div class='output co'>#&gt; <span class='message'>Finally, select 2 after passing threshold 0.1</span></div><div class='output co'>#&gt; <span class='message'>===================</span></div><div class='output co'>#&gt; <span class='message'>Fitting feature: osCN</span></div><div class='output co'>#&gt; 2 : *
#&gt; 3 : *
#&gt; 4 : *
#&gt; 5 : *
#&gt; 6 : *
#&gt; 7 : *
#&gt; 8 : *
#&gt; 9 : *
#&gt; 10 : *
#&gt; 11 : *
#&gt; 12 : *
#&gt; 13 : *
#&gt; 14 : *
#&gt; 15 : *</div><div class='output co'>#&gt; <span class='message'>Select 2 according to BIC</span></div><div class='output co'>#&gt; <span class='message'>Finally, select 2 after passing threshold 0.1</span></div><div class='output co'>#&gt; <span class='message'>===================</span></div><div class='output co'>#&gt; <span class='message'>Fitting feature: bpchrarm</span></div><div class='output co'>#&gt; 2 : *
#&gt; 3 : *
#&gt; 4 : *
#&gt; 5 : *
#&gt; 6 : *
#&gt; 7 : *
#&gt; 8 : *
#&gt; 9 : *
#&gt; 10 : *
#&gt; 11 : *
#&gt; 12 : *
#&gt; 13 : *
#&gt; 14 : *
#&gt; 15 : *</div><div class='output co'>#&gt; <span class='message'>Select 2 according to BIC</span></div><div class='output co'>#&gt; <span class='message'>Finally, select 2 after passing threshold 0.1</span></div><div class='output co'>#&gt; <span class='message'>===================</span></div><div class='output co'>#&gt; <span class='message'>Fitting feature: changepoint</span></div><div class='output co'>#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     25 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     19 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     14 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     14 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     13 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     15 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     12 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     14 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     15 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     20 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     15 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     14 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     14 Log-likelihood: NaN
#&gt; 2 : *
#&gt; 3 : *
#&gt; 4 : *
#&gt; 5 : *
#&gt; 6 : *
#&gt; 7 : *
#&gt; 8 : *
#&gt; 9 : *
#&gt; 10 : *
#&gt; 11 : *
#&gt; 12 : *
#&gt; 13 : *
#&gt; 14 : *
#&gt; 15 : *</div><div class='output co'>#&gt; <span class='message'>Fitting feature: copynumber</span></div><div class='output co'>#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     12 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     10 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;      9 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     12 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     11 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     10 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     10 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     12 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     10 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     10 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     11 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     11 Log-likelihood: NaN
#&gt; Error in FLXfit(model = model, concomitant = concomitant, control = control,  : 
#&gt;     10 Log-likelihood: NaN
#&gt; 2 : *
#&gt; 3 : *
#&gt; 4 : *
#&gt; 5 : *
#&gt; 6 : *
#&gt; 7 : *
#&gt; 8 : *
#&gt; 9 : *
#&gt; 10 : *
#&gt; 11 : *
#&gt; 12 : *
#&gt; 13 : *
#&gt; 14 : *
#&gt; 15 : *</div><div class='output co'>#&gt; <span class='message'>=&gt; Step: calculating the sum of posterior probabilities</span></div><div class='output co'>#&gt; <span class='message'>Calculating the sum of posterior probabilities for bp10MB</span></div><div class='output co'>#&gt; <span class='message'>Calculating the sum of posterior probabilities for copynumber</span></div><div class='output co'>#&gt; <span class='message'>Calculating the sum of posterior probabilities for changepoint</span></div><div class='output co'>#&gt; <span class='message'>Calculating the sum of posterior probabilities for bpchrarm</span></div><div class='output co'>#&gt; <span class='message'>Calculating the sum of posterior probabilities for osCN</span></div><div class='output co'>#&gt; <span class='message'>Calculating the sum of posterior probabilities for segsize</span></div><div class='output co'>#&gt; <span class='message'>=&gt; Done.</span></div><div class='input'><span class='co'># }</span>
<span class='co'># Use method designed by Wang, Shixiang et al.</span>
<span class='no'>cn_tally_W</span> <span class='kw'>&lt;-</span> <span class='fu'>sig_tally</span>(<span class='no'>cn</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"W"</span>)</div><div class='output co'>#&gt; <span class='message'>=&gt; Step: getting copy number features</span></div><div class='output co'>#&gt; <span class='message'>Getting breakpoint count per 10 Mb...</span></div><div class='output co'>#&gt; <span class='message'>Getting breakpoint count per chromosome arm...</span></div><div class='output co'>#&gt; <span class='message'>Getting copy number...</span></div><div class='output co'>#&gt; <span class='message'>Getting change-point copy number change...</span></div><div class='output co'>#&gt; <span class='message'>Getting length of chains of oscillating copy number...</span></div><div class='output co'>#&gt; <span class='message'>Getting (log10 based) segment size...</span></div><div class='output co'>#&gt; <span class='message'>Getting the minimal number of chromosome with 50% CNV...</span></div><div class='output co'>#&gt; <span class='message'>Getting burden of chromosome...</span></div><div class='output co'>#&gt; <span class='message'>=&gt; Step: generating copy number components</span></div><div class='output co'>#&gt; <span class='message'>=&gt; Step: generating components by sample matrix</span></div><div class='output co'>#&gt; <span class='message'>=&gt; Done.</span></div><div class='input'><span class='co'># \donttest{</span>
<span class='co'># Prepare SBS signature analysis</span>
<span class='no'>laml.maf</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"tcga_laml.maf.gz"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"maftools"</span>)
<span class='no'>laml</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='read_maf.html'>read_maf</a></span>(<span class='kw'>maf</span> <span class='kw'>=</span> <span class='no'>laml.maf</span>)</div><div class='output co'>#&gt; -Reading
#&gt; -Validating
#&gt; -Silent variants: 475 
#&gt; -Summarizing
#&gt; -Processing clinical data
#&gt; --Missing clinical data
#&gt; -Finished in 0.330s elapsed (0.250s cpu) </div><div class='input'><span class='kw'>if</span> (<span class='fu'><a href='https://rdrr.io/r/base/library.html'>require</a></span>(<span class='st'>"BSgenome.Hsapiens.UCSC.hg19"</span>)) {
  <span class='no'>mt_tally</span> <span class='kw'>&lt;-</span> <span class='fu'>sig_tally</span>(
    <span class='no'>laml</span>,
    <span class='kw'>ref_genome</span> <span class='kw'>=</span> <span class='st'>"BSgenome.Hsapiens.UCSC.hg19"</span>,
    <span class='kw'>prefix</span> <span class='kw'>=</span> <span class='st'>"chr"</span>, <span class='kw'>add</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>useSyn</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
  )
} <span class='kw'>else</span> {
  <span class='fu'><a href='https://rdrr.io/r/base/message.html'>message</a></span>(<span class='st'>"Please install package 'BSgenome.Hsapiens.UCSC.hg19' firstly!"</span>)
}</div><div class='output co'>#&gt; <span class='message'>Loading required package: BSgenome.Hsapiens.UCSC.hg19</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: BSgenome</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: S4Vectors</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: stats4</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: 'S4Vectors'</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from 'package:future':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    values</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from 'package:NMF':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    nrun</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from 'package:pkgmaker':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    new2</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from 'package:base':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    expand.grid</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: IRanges</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: 'IRanges'</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from 'package:purrr':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    reduce</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from 'package:grDevices':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    windows</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: GenomeInfoDb</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: GenomicRanges</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: Biostrings</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: XVector</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: 'XVector'</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from 'package:purrr':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    compact</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: 'Biostrings'</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from 'package:base':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    strsplit</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: rtracklayer</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Chromosome names in MAF must match chromosome names in reference genome.</span>
#&gt; <span class='warning'>Ignorinig 101 single nucleotide variants from missing chromosomes chr23</span></div><div class='output co'>#&gt; -Extracting 5' and 3' adjacent bases
#&gt; -Extracting +/- 20bp around mutated bases for background C&gt;T estimation
#&gt; -Estimating APOBEC enrichment scores
#&gt; --Performing one-way Fisher's test for APOBEC enrichment
#&gt; ---APOBEC related mutations are enriched in  3.315 % of samples (APOBEC enrichment score &gt; 2 ;  6  of  181  samples)
#&gt; -Creating mutation matrix
#&gt; --matrix of dimension 188x96</div><div class='input'># }
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#methods-by-class-">Methods (by class)</a></li>
      <li><a href="#references">References</a></li>
      <li><a href="#see-also">See also</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Shixiang Wang</p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://orcid.org/0000-0001-9855-7357'>Shixiang Wang</a>, <a href='https://orcid.org/0000-0002-7736-0077'>Xue-Song Liu</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


