<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>sigminer: Capture Genomic Variation Signatures using Non-Negative Matrix Factorization — sigminer • sigminer</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="sigminer: Capture Genomic Variation Signatures using Non-Negative Matrix Factorization — sigminer" />

<meta property="og:description" content="sigminer provides functions for identification of
copy number signatures (Geoffrey et al. (2018) doi: 10.1038/s41588-018-0179-8
) and
mutational signatures (Alexandrov et al. (2013) doi: 10.1038/nature12477
)
by non-negative matrix factorization, signature analysis and visualization.
It can be used to capture signatures of genomic variation, compare genotype or
phenotype features of different signatures and thus uncover the relationship
between the mechanism of genomic variation and phenotypes in cancer." />

<meta property="og:image" content="/logo.png" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sigminer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/sigminer.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ShixiangWang/sigminer">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>sigminer: Capture Genomic Variation Signatures using Non-Negative Matrix Factorization</h1>
    <small class="dont-index">Source: <a href='https://github.com/ShixiangWang/sigminer/blob/master/R/sigminer.R'><code>R/sigminer.R</code></a></small>
    <div class="hidden name"><code>sigminer.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>sigminer provides functions for identification of
copy number signatures (Geoffrey et al. (2018) doi: <a href='http://doi.org/10.1038/s41588-018-0179-8'>10.1038/s41588-018-0179-8</a>
) and
mutational signatures (Alexandrov et al. (2013) doi: <a href='http://doi.org/10.1038/nature12477'>10.1038/nature12477</a>
)
by non-negative matrix factorization, signature analysis and visualization.
It can be used to capture signatures of genomic variation, compare genotype or
phenotype features of different signatures and thus uncover the relationship
between the mechanism of genomic variation and phenotypes in cancer.</p>
    
    </div>

        
    <h2 class="hasAnchor" id="read-functions"><a class="anchor" href="#read-functions"></a>read_ functions</h2>

    
    <p>Read data into R objects:</p><ul>
<li><p><code><a href='read_copynumber.html'>read_copynumber()</a></code> - read <code>data.frame</code>/files into <a href='CopyNumber-class.html'>CopyNumber</a> object.</p></li>
<li><p><code><a href='read_maf.html'>read_maf()</a></code> - read <code>data.frame</code>/file into <a href='MAF-class.html'>MAF</a> object. This is powered by
<code><a href='https://www.rdocumentation.org/packages/maftools/topics/read.maf'>read.maf</a></code> function in <strong>maftools</strong> package.</p></li>
<li><p><code><a href='read_variation.html'>read_variation()</a></code> - combine a <a href='CopyNumber-class.html'>CopyNumber</a> object and a <a href='MAF-class.html'>MAF</a> object as
a <a href='GenomicVariation-class.html'>GenomicVariation</a> object. <strong>It is not useful for now</strong>.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="sig-functions"><a class="anchor" href="#sig-functions"></a>sig_ functions</h2>

    
    <p>Signature analysis for genomic variations:</p><ul>
<li><p><code><a href='sig_prepare.html'>sig_prepare()</a></code> - do preparation step of signature analysis for <a href='CopyNumber-class.html'>CopyNumber</a>
or <a href='MAF-class.html'>MAF</a> object.</p></li>
<li><p><code><a href='sig_estimate.html'>sig_estimate()</a></code> - provides survey plot and consensus map to user
for selecting the best signature number.</p></li>
<li><p><code><a href='sig_extract.html'>sig_extract()</a></code> - extract signatures based on specified signature number.</p></li>
<li><p><code><a href='sig_assign_samples.html'>sig_assign_samples()</a></code> - assign samples to signatures. This classify samples
into different subgroups based on the dominant signature.</p></li>
<li><p><code><a href='sig_get_activity.html'>sig_get_activity()</a></code> - obtain signature activity in samples.</p></li>
<li><p><code><a href='sig_get_correlation.html'>sig_get_correlation()</a></code> - obtain correlation matrix between signatures activity.</p></li>
<li><p><code><a href='sig_get_similarity.html'>sig_get_similarity()</a></code> - get similarity between signatures.</p></li>
<li><p><code><a href='sig_summarize_subtypes.html'>sig_summarize_subtypes()</a></code> - get summary of signature subtypes.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="draw-functions"><a class="anchor" href="#draw-functions"></a>draw_ functions</h2>

    
    <p>Result visualization for copy number data:</p><ul>
<li><p><code><a href='draw_cn_distribution.html'>draw_cn_distribution()</a></code> - plot copy number distribution either by length or chromosome.</p></li>
<li><p><code><a href='draw_cn_features.html'>draw_cn_features()</a></code> - plot copy number feature distribution.</p></li>
<li><p><code><a href='draw_cn_components.html'>draw_cn_components()</a></code> - plot mixture fit model components.</p></li>
</ul>
    <p>Result visualization for signature analysis:</p><ul>
<li><p><code><a href='draw_sig_profile.html'>draw_sig_profile()</a></code> - plot signature profile.</p></li>
<li><p><code><a href='draw_sig_activity.html'>draw_sig_activity()</a></code> - plot signature activity.</p></li>
<li><p><code><a href='draw_sig_corrplot.html'>draw_sig_corrplot()</a></code> - plot correlation between signature activities.</p></li>
<li><p><code><a href='draw_subtypes_comparison.html'>draw_subtypes_comparison()</a></code> - plot comparison between signature subtypes.</p></li>
</ul>
    <p>Result visualization for <a href='MAF-class.html'>MAF</a> is provide by <strong>maftools</strong> package,
please read its <a href='https://bioconductor.org/packages/release/bioc/vignettes/maftools/inst/doc/maftools.html'>vignette</a>.</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      
      <li><a href="#read-functions">read_ functions</a></li>

      <li><a href="#sig-functions">sig_ functions</a></li>

      <li><a href="#draw-functions">draw_ functions</a></li>
          </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Shixiang Wang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

